File: services/ordering/src/main/java/io/wkrzywiec/fooddelivery/ordering/domain/OrderingEvent.java
package io.wkrzywiec.fooddelivery.ordering.domain;

import io.wkrzywiec.fooddelivery.commons.infra.store.DomainEvent;

import java.math.BigDecimal;
import java.util.List;

public interface OrderingEvent extends DomainEvent {

    String orderId();

    @Override
    default String streamId() {
        return orderId();
    }

    record OrderCreated(
            String orderId,
            int version,
            String customerId,
            String farmId,
            String address,
            List<Item> items,
            BigDecimal deliveryCharge,
            BigDecimal total
    ) implements OrderingEvent {}

    record OrderCanceled(
            String orderId,
            int version,
            String reason
    ) implements OrderingEvent {}

    record OrderInProgress(
            String orderId,
            int version
    ) implements OrderingEvent {}

    record TipAddedToOrder(
            String orderId,
            int version,
            BigDecimal tip,
            BigDecimal total
    ) implements OrderingEvent {}

    record OrderCompleted(
            String orderId,
            int version
    ) implements OrderingEvent {}
}

File: services/ordering/src/main/kotlin/io/wkrzywiec/fooddelivery/ordering/domain/OrderingEvent.kt
package io.wkrzywiec.fooddelivery.ordering.domain

import io.wkrzywiec.fooddelivery.commons.infra.store.DomainEvent
import java.math.BigDecimal
import java.util.*

interface OrderingEvent : DomainEvent {
    fun orderId(): UUID

    override fun streamId(): UUID {
        return orderId()
    }

    data class OrderCreated(
        @JvmField val orderId: UUID,
        @JvmField val version: Int,
        @JvmField val customerId: String,
        @JvmField val farmId: String,
        @JvmField val address: String,
        @JvmField val items: List<Item>,
        @JvmField val deliveryCharge: BigDecimal,
        @JvmField val total: BigDecimal
    ) : OrderingEvent {
        override fun orderId(): UUID = orderId
        override fun version(): Int = version
    }

    data class OrderCanceled(
        @JvmField val orderId: UUID,
        @JvmField val version: Int,
        @JvmField val reason: String
    ) : OrderingEvent {
        override fun orderId(): UUID = orderId
        override fun version(): Int = version
    }

    data class OrderInProgress(
        @JvmField val orderId: UUID,
        @JvmField val version: Int
    ) : OrderingEvent {
        override fun orderId(): UUID = orderId
        override fun version(): Int = version
    }

    data class TipAddedToOrder(
        @JvmField val orderId: UUID,
        @JvmField val version: Int,
        @JvmField val tip: BigDecimal,
        @JvmField val total: BigDecimal
    ) : OrderingEvent {
        override fun orderId(): UUID = orderId
        override fun version(): Int = version
    }
    
    data class OrderCompleted(
        @JvmField val orderId: UUID,
        @JvmField val version: Int
    ) : OrderingEvent {
        override fun orderId(): UUID = orderId
        override fun version(): Int = version
    }
}
